cmake: drop `fixCmakeFiles` hook

emilazy opened the pull request:
Stacked on top of https://github.com/NixOS/nixpkgs/pull/394444; only the additional commits here are relevant. This is based on the patch I’ve built a ton of stuff on Darwin with (albeit with the sandbox enabled), but not identical, so this is currently completely untested. See commit messages for details.

Closes https://github.com/NixOS/nixpkgs/issues/24215.

Closes https://github.com/NixOS/nixpkgs/pull/232522.
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy committed:
cmake: split search path patch

To be honest, I’m not sure exactly how useful the libc one is or
why it doesn’t need to include the C++ library and the like, but
splitting this up helps with the following work.

emilazy committed:
cmake: fold `fixCmakeFiles` into the search path patch

A verbatim patch representing its changes to CMake 4.1.1 provides a
useful base for further modifications as we drop the hook.

emilazy committed:
cmake: undo `fixCmakeFiles` Darwin changes from `postFixup`

emilazy committed:
cmake: undo `fixCmakeFiles` changes in tests

This is a huge amount of diff noise for things that aren’t installed
and that we don’t even run. If anyone wants to try getting the tests
running, they can fix reproducibility problems that may be addressed
by these changes in a more targeted manner.

emilazy committed:
cmake: undo `fixCmakeFiles` changes in release engineering files

We don’t run CMake’s CI or its release scripts.

emilazy committed:
cmake: undo `fixCmakeFiles` changes in documentation and comments

In some cases these may make the statements more accurate to our
patched version, but in practice they just add lines to the diff and
often make it more confusing or outright wrong instead. I don’t
think we even install the documentation, anyway.

emilazy committed:
cmake: undo `fixCmakeFiles` changes in CPack

I don’t think these were correct.

emilazy committed:
cmake: undo `fixCmakeFiles` MSYS changes

If we end up supporting MSYS in Nixpkgs, it would make more sense to
address this at the root in `FindMSYS.cmake` instead of adding this
diff noise.

emilazy committed:
cmake: undo `fixCmakeFiles` Git for Windows changes

Again, it makes more sense to address this systematically as part
of Windows support, especially as some of the paths weren’t even
being mangled.

emilazy committed:
cmake: undo `fixCmakeFiles` SunPro changes

I think this will make the behaviour more correct? For people using
the SunPro compiler with Nixpkgs? A thing that will definitely happen?

emilazy committed:
cmake: undo `fixCmakeFiles` Android NDK changes

I believe that these changes are incorrect because the NDK operates
with paths relative to a sysroot.

emilazy committed:
cmake: undo `fixCmakeFiles` obscure platform changes

These are just not worth the diff noise, and we have no way to verify
them for correctness.

emilazy committed:
cmake: undo `fixCmakeFiles` installation path changes

We set these paths ourselves anyway, and the patched logic makes
no sense.

emilazy committed:
cmake: undo `fixCmakeFiles` system library determination changes

This won’t apply to Nixpkgs use anyway, and the patched logic made
no sense.

emilazy committed:
cmake: undo more `fixCmakeFiles` Darwin changes

These are as incorrect as the earlier ones, but were not previously
being reverted.

emilazy committed:
cmake: patch out more impure Darwin paths

`fixCmakeFiles` didn’t catch these.

emilazy committed:
cmake: patch out guesses based on distribution release files

emilazy committed:
cmake: patch out search for system curl CA bundle files

emilazy committed:
cmake: adjust patched `FindDCMTK.cmake` logic

No idea what this thing even is, frankly.

emilazy committed:
cmake: drop code referencing `/var/empty` from the search path patch

These are almost all dead code – some references to `/bin` and
`/sbin` and some Windows paths notwithstanding – and dropping them
entirely makes the patch shorter and more effective.

It is likely that a lot of this patch is still superfluous; we
don’t really care about the purity of unsandboxed builds on Linux,
for instance, so any code path that is never taken on Darwin could
probably be dropped. But this provides a conservative base to drop the
problematic hook without compromise for CMake‐distributed modules. Of
course, the long‐term solution should be proper sandboxing on macOS.

emilazy committed:
cmake: drop `fixCmakeFiles` hook

This regularly broke things – as can be seen from the number
of packages that were opting out of it – while providing only a
dubious and partial benefit to purity that other build systems did
not “enjoy”.

It has become associated with Darwin, but was actually introduced
in 2007 as part of 67d41efde183d84f967e350e3e75d8f8dedd165d,
a work‐in‐progress commit for KDE 4 that introduced the CMake
setup hook. It has haunted us ever since, and will presumably become
even better at doing so now that it is a dead ghost.

Of course, it is possible and even likely that more impure paths will
leak into Darwin builds from third‐party CMake modules included
in packages. However, considering it already disregarded many of the
most common macOS system paths, it will be better overall to address
these in a targeted manner than dealing with the effects of this
brute‐force hook. In the long run, a good sandbox is going to be
the solution for reliable build reproducibility on the platform.

K900 commented:
[Full graphical ISO](https://static.0upti.me/iso-cmake4.iso) built on top of [my integration branch](https://github.com/K900/nixpkgs/tree/cmake4-integration).

emilazy commented:
> This is impressive work. What sort of feedback and testing are you looking for, @emilazy?

Uh, good question.

As I say in the commit messages, the hook was introduced back in the mists of time for some Linux KDE 4 stuff, probably before there was any build sandboxing. Nominally the current benefit of the hook – other than “just dropping it broke the build of CMake itself a while back” – is “avoid stuff from the system leaking into development shells and non‐sandboxed builds on Darwin and non‐NixOS Linux”.
That’s pretty much a pervasive problem with all build systems, but if you want to justify it, you can say that CMake modules seem to like to hardcode paths more often than Autotools and Meson builds seem to. A lot of them are out‐of-tree and so we can’t do anything about them if we’re dropping the hook other than patching one‐by‐one (perhaps with an opt‐in hook for repeat offenders, or at least having a hook that is more scoped to detect and patch specific modules), but we can at least patch the ones shipped with CMake itself.

The *downside* of the hook is that it it’s a gross mass regular expression replacement that mangles paths, including things like `./src/opt/stands/for/optimized`, or seemingly “absolute” paths on Darwin that are actually load‐bearing and relative to the SDK. So that’s why we really want to get rid of it, and I err on the side of only dealing with purity issues beyond the CMake package itself once they observably become a problem – proper sandboxing is the only true fix for those, anyway. I suspect that most of what remains in the patch that replaces the hook is basically unnecessary and that over time we can cut it down significantly, since preventing leaks of the system into development shells is kind of impossible and Darwin needs a sandbox good enough to enable by default. But this is intended as a conservative change for now.

We’ve collectively built a lot of stuff with this on Linux and (on a previous version of the patch) Darwin, so there’s been a reasonable level of practical validation. OTOH, my Darwin builds have been sandboxed, so we’ve had relatively little detection of potential purity issues.

The kind of review I’d ideally hope for would require CMake expertise that I’m not sure we have on tap. We’re probably just going to merge this within the next day regardless, so basic cross‐checking is the most valuable thing. https://github.com/NixOS/nixpkgs/pull/442167/commits/dda6c95d519d120c739ff0ca215efb41a7bbf698 and https://github.com/NixOS/nixpkgs/pull/442167/commits/1ade29ef68e4f17c4ecc7293842d96444ee983e5 can be ignored as being huge patches that are entirely rote and mechanical, unless someone really feels like reproducing them; the other commits here involve some judgement calls and rearranging of logic, so could use looking over. In descending order of priority, something like:

1. Does the final state of the patch have any changes that look obviously bad?

2. Are there any cases where the changes to the patch in a commit don’t seem to obviously line up with what the commit says it does?

3. 5. Does it make sense to drop the call to `brew` to find the Homebrew installation path in https://github.com/NixOS/nixpkgs/pull/442167/commits/28b88d869fd08ce86cc5565e8930fa8dab428c46, but leave in the `xcode-select -print-path` call so that the Nixpkgs CMake can find the native compiler tools where relevant? Neither of them matter inside a Nix build – even unsandboxed – unless `/usr/bin` or `/usr/local/bin` or `/opt/homebrew/bin` get added to the executable search path, but I don’t know what the right thing is here for development shells, which I haven’t tested at all.

   (In writing this, I think I have talked myself out of this inconsistency, so I might change it before merge. I don’t know whether “if `brew` is on the `$PATH` you can have Homebrew too, it’s fine” or “no looking at the system stuff ever” is the right call. The latter makes our behaviour diverge from upstream CMake outside of a Nix development shell for the sake of making those more pure. Maybe the best solution here is to check for the relevant Nix environment variable and adjust behaviour accordingly, but then that’s more moving parts to the patch. I’ll probably just drop the `xcode-select` call and we can revisit this later. But this is the kind of thing it’d be especially helpful to have feedback on, particularly from people who use Nixpkgs CMake on Darwin in development shells or outside of Nix environments entirely.)

4. Could the CPack changes in https://github.com/NixOS/nixpkgs/pull/442167/commits/71ce6702e2b042f9bc990827c576f0bbf01e57fe actually cause any issues, given that IIRC some packages use CPack as part of their build/installation process? Is my idea of what they’re being used for accurate? Similarly for https://github.com/NixOS/nixpkgs/pull/442167/commits/33adeef8411fefdcfbf86f215f980fdf3b53a98b and https://github.com/NixOS/nixpkgs/pull/442167/commits/d2f96c57bcb22b0a166219896452b86fc8ccc86f.

5. Does https://github.com/NixOS/nixpkgs/pull/442167/commits/bf5e08f25ea4abf05a216802c19422d36b8dfdca make sense? There are lots of ways to detect the distribution version. We expose the `sw_vers` results on macOS. The cases where it sniffs in ways that could affect development shells on non‐NixOS Linux seem weird, though. I dunno.

6. Is https://github.com/NixOS/nixpkgs/pull/442167/commits/c57db6dbdb8a29284cb3aec5244b3dc36623dcc3 going to break TLS stuff? It changes behaviour. I think some of those paths exist sometimes inside the sandbox on Linux. I believe curl will find the CA certificates fine by itself, though…? But I don’t know exactly how all this stuff wires up together in CMake.

7. Am I right about https://github.com/NixOS/nixpkgs/pull/442167/commits/45039c1b7b73d2019d360a956cb30f1336f0fc97? How does the NDK sysroot stuff even work with our Android build environment? (cc @numinit?)

8. Do we have any hope of convincing CMake upstream to just do less of this kind of path‐guessing thing in general?

(Diffs‐of‐diffs are kind of weird. You can manually re‐apply these patches on top of CMake 4.1.1 if you want to see the changes without getting confused by green lines that are actually getting deleted.)

Anyway, probably we will just actually find out if any of this stuff is problematic once it reaches users, in the end. The patch will at least be more legible about what’s actually being changed. But I’d still appreciate any review and feedback in the limited window we have before it’s going to merge.

emilazy committed:
cmake: add `sysctl(8)` to the Darwin/BSD paths patch

Used for determining core count and Apple Silicon stuff.

Also drop unnecessary `@ps@` replacements in `#elif` branches that
aren’t taken, and correct a comment about `ps(1)` being in the Linux
`stdenv`, since… it does not appear to be.

philiptaron commented on pkgs/by-name/cm/cmake/package.nix:
```diff
@@ -60,24 +61,30 @@ stdenv.mkDerivation (finalAttrs: {
-    replaceVars ./darwin-bsd-ps-abspath.patch {
+  ++ lib.optionals (stdenv.hostPlatform.isDarwin || stdenv.hostPlatform.isFreeBSD) [
+    (replaceVars ./darwin-bsd-binary-paths.patch {
+      # `ps(1)` is theoretically used on Linux too, but only when
+      # `/proc` is inaccessible, so we can skip the dependency.
```
Yeah, that makes sense.

philiptaron reviewed (approved):
I believe.
